### PR TITLE
feat: readonly checkbox

### LINKED ISSUE #12243: Read-only Inputs: Checkbox Implementation
### Reference #12241 

image

image

#### Acceptance Criteria

- [ ] Coded component in `/package/react`
- [ ] Add a prop for read-only, add styles for said prop
- [ ] Controls for each state in playground
- [ ] Default story for easy copy and paste
- [ ] RTL tests for the read-only class on the prop

### PR DESCRIPTION
**Pull request by lee-chase**
Contributes to #2177
Closes #12243

Adds the read-only feature to the Checkbox component

#### Changelog

**Added**

- Adds readOnly property to Checkbox and functionality
- Adds test for read-only checkbox
- Adds styling for read-only checkbox (checked and unchecked).

#### Testing / Reviewing

Reviewed the result in Storybook and a added read-only test.

### CODE REVIEW

File: packages/react/src/components/Checkbox/Checkbox.js [hunk 1]
```diff
@@ -19,6 +19,7 @@ const Checkbox = React.forwardRef(function Checkbox(
     onChange,
     indeterminate,
     hideLabel,
+    readOnly,
```
Comment (tay1orjones) [inline packages/react/src/components/Checkbox/Checkbox.js]: Even though checkboxes don't support the standard attribute `readonly`, what are your thoughts on naming this `readonly` (lowercase) for consistency with the other inputs that will use `readonly`?

File: packages/react/src/components/Checkbox/Checkbox.js [hunk 2]
```diff
@@ -54,6 +60,14 @@ const Checkbox = React.forwardRef(function Checkbox(
...
           }
         }}
+        // readonly attribute not applicable to type="checkbox"
+        // see - https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox
+        aria-readonly={readOnly}
+        onClick={(evt) => {
+          if (readOnly) {
+            evt.preventDefault();
+          }
+        }}
```
Comment (tay1orjones) [inline packages/react/src/components/Checkbox/Checkbox.js]: Is there any potential use case where click should still be fired, even if it's readonly? Maybe the user needs to be shown a notification or something else to indicate why it's readonly?

File: packages/react/src/components/Checkbox/__tests__/Checkbox-test.js [hunk 3]
```diff
@@ -78,10 +78,20 @@ describe('Checkbox', () => {
     );
   });
 
-  describe('indeterminate', () => {
-    it('should set the indeterminate attribute of the <input> to true', () => {
-      render(<Checkbox id="test" indeterminate labelText="test-label" />);
-      expect(screen.getByRole('checkbox').indeterminate).toBe(true);
-    });
+  it('should NOT call the `onChange` prop when readonly', () => {
```
Comment (tay1orjones) [inline packages/react/src/components/Checkbox/__tests__/Checkbox-test.js:81]: If onClick is disabled as well, this test duplicated but for `onClick` would be nice

File: packages/styles/scss/components/checkbox/_checkbox.scss [hunk 4]
```diff
@@ -180,6 +180,31 @@
     background-color: $icon-disabled;
   }
 
+  //----------------------------------------------
+  // Read-only
+  // ---------------------------------------------
+  // should this be a token of it's own?
+  $border-readonly: $icon-disabled;
```
Comment (tay1orjones) [inline packages/styles/scss/components/checkbox/_checkbox.scss]: I'd just pull this out for now unless there's clear indication from design that we're going to add tokens for this.

**Review by tay1orjones — COMMENTED** [review]
Overall looks good, just a few comments

File: packages/react/src/components/Checkbox/Checkbox.js (hunk 1 above)
Comment (lee-chase) [reply packages/react/src/components/Checkbox/Checkbox.js]: My preference would be to use lowercase `readonly` everywhere as we are using it for the standards defined purpose. This is simply repeating the pattern used in TextInput for consistency.

File: packages/react/src/components/Checkbox/Checkbox.js (hunk 2 above)
Comment (lee-chase) [reply packages/react/src/components/Checkbox/Checkbox.js]: Agreed. I believe I assumed `preventDefault` would not stop propagation.

Added test for onClick and passed on if the user supplies a method.

Comment (lee-chase) [reply packages/react/src/components/Checkbox/Checkbox.js]: Note: onClick is normally passed through from read-only inputs.

File: packages/react/src/components/Checkbox/__tests__/Checkbox-test.js (hunk 3 above)
Comment (lee-chase) [reply packages/react/src/components/Checkbox/__tests__/Checkbox-test.js:81]: Added a check for onClick being propagated.

File: packages/styles/scss/components/checkbox/_checkbox.scss (hunk 4 above)
Comment (tay1orjones) [reply packages/styles/scss/components/checkbox/_checkbox.scss]: Confirmation to pull this out for now - https://github.com/carbon-design-system/carbon/issues/12243#issuecomment-1293607905:

> I personally don't think we need read-only tokens. I think using a mix of regular and disabled color tokens in this context is acceptable. If the style ever diverges, we can make a specific one. Its only used in a few read-only components. We could make a whole suite of read-only tokens but I think that is overkill.

**Review by aagonzales — CHANGES_REQUESTED** [review]
The styles look good but there are some interaction problems.

I'm still able to click on the check-box, has a cursor change to a hand and focus border. My assumption is those things shouldn't be possible because it is giving the illusion that its interactive. (Interactions aren't explicitly stated in the spec though we might need to circle back with Devin and team). 

Update: Oh is that the `onClick` thing you're talking about with @tay1orjones? At a minimum, can we not have the cursor change, can it just remain an arrow or switch to the `not-allowed` cursor? 

Oct-28-2022 14-03-00

**Comment by lee-chase** [comment]
@aagonzales @tay1orjones was referring to the internals of the component I believe.

With regards to typical behaviour `readonly` is only applied to text input fields. The only distinction from disabled is that it is focusable and will be submitted with forms. 

With regards to the cursor I'm not sure if https://github.com/quarryboy or Mike Eaker would care to comment? 

The default behaviour (for input fields) can be seen in the link below. The cursor stays as it would normally. That is to say it displays the `text` cursor in most cases, the `deafault` pointer for date and the `se-resize` at the corner of a textarea.

 https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/readonly

**Comment by aagonzales** [comment]
@lee-chase I talked with Mike and Devin over Slack. Here are the confirmed expected interactive behaviors. So the changes needed are:
- [ ] Hovering over the checkbox icon there should be no cursor change (remains as an `arrow`)
- [ ] Hovering over the checkbox labels the cursor should change to a `type` cursor (not a `hand`)

image

**Comment by lee-chase** [comment]
@tw15egan merge conflict fixed

**Comment by lee-chase** [comment]
@aagonzales addressed cursor issue

File: packages/react/src/components/Checkbox/Checkbox.js (hunk 1 above)
Comment (lee-chase) [reply packages/react/src/components/Checkbox/Checkbox.js]: Unfortunately React says not to use 'readonly' https://reactjs.org/docs/dom-elements.html#all-supported-html-attributes

**Review by aagonzales — APPROVED** [review]
Visually looks good, matches spec 👍

**Review by alisonjoseph — APPROVED** [review]

**Review by tw15egan — APPROVED** [review]

**Pull request merged**